FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+say you are inserting a new object into the database. This is best done by giving your
+target table a primary key column that is a UUID. Then you can run a sequence of steps
+something like this:
+ 1. Step that generates a random UUID and stores it in the working map; _OR_ hand a random
+ UUID into the flight as an input parameter. If this step is run more than once, it will
+ simply generate a different UUID.
+ 1. Step that runs a database transaction. The database transaction does:
+    1. Read the table to see if a row with the UUID already exists. If so, return
+    success. A previous run of the step inserted that row.
+    1. Insert the row into the table.
+    If this step is run more than once, either the database transaction will have rolled
+    back and the row will be re-inserted, or the database transaction will have committed
+    and the read will detect that the insert already happened successfully.
+
+
+A variation on this approach is for the step to trap the duplicate key exception from the
+database system and consider that a success.
+
+### Existence Checking <a name="existencechecking"></a>
+The database transaction case brings up the general problem of coordinating existence
+checking and idempotent steps. There are a few common pitfalls.
+
+#### Pitfall 1: deleting someone else's object
+Imagine that you want to create a thing named "foo". The thing might be a file, a cloud
+object, a database row, whatever. You write a step like this:
+
+```
+Step: Create a thing
+ do():
+   create "foo"
+ undo():
+   delete "foo"
+```
+But what if "foo" already exists when this step is run? The execution will go:
+ 1. Stairway runs `do()` and call create "foo". That results in a "foo already exists"
+ exception.
+ 1. Stairway catches the exception and decides the flight is failed. It
+ transitions to undoing.
+ 1. Stairway runs `undo()` and calls delete "foo". It deletes the pre-existing object.
+
+#### Pitfall 2: failing when you already succeeded
+
+```
+Step: Create a thing - take 2
+ do():
+   if "foo" exists {
+     write "foo already existed" to the working map
+     throw "foo exists"
+   }
+   create "foo"
+ undo():
+   if the working map does says "foo already existed" {
+     then return success
+   }
+   delete "foo"
+```
+That is better. When Stairway catches the exception and transitions to undoing, it writes
+the flight state, so the working map gets written. The undo can skip the delete. Yay!
+
+The problem is that the `do()` is no longer idempotent. Suppose there is a server failure right
+after the "foo" thing is created, but before Stairway records the step result. On
+recovery, Stairway will rerun the `do()` step. It will say, "foo already existed", and the
+flight will fail.
+
+#### Pitfall 3: concurrent operation
+
+A more reliable way to accomplish the existence check is to do it in a separate step. So
+now you have this pseudo-code:
+```
+Step 1: Check thing existence
+ do():
```

Comment:
This sounds like what we have in PF-589 where we decided there was another race condition if two flights try to create the same thing simultaneously.